### PR TITLE
Feat/favorites api

### PR DESCRIPTION
**Pull request by cbeach47**
##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] tests are included

##### Description of change
Build out of the `portlet-list` API endpoint feature.

Tests to support this feature is in https://github.com/uPortal-Project/uPortal-start/pull/575

Future efforts:
- Setup the portlet-list portlet entries to be a foreign key constraint on portlet definition fnames.
- Include 'details' parameter in portlet-list retrieval to inflate portlets (removes the need for a second API call)
- Paginate retrieval of all portlet-lists
- Gradle task to migrate old favorites to portlet list favorites

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

**Comment by cbeach47** [comment]
I don't expect the overall structure to change - possibly tweaks as I finish out the integration tests.

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/portletlist/PortletListRESTController.java [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+        } catch (Exception e) {
+            log.warn("User [{}] tried to create a portlet-list with bad json - {}", person.getUserName(), e.getMessage());
+            return prepareResponse(response, null, "Unparsable portlet-list JSON", HttpServletResponse.SC_BAD_REQUEST);
+        }
+
+        try {
+            final IPortletList created = portletListService.createPortletList(
+                person, input);
+            response.setHeader("Location", created.getId());
```
[the remark this answers is not in the document]
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/portletlist/PortletListRESTController.java:157]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/portletlist/PortletListRESTController.java:157]: Not a concern - we are setting the id field from hibernate's generated ID, not from user input.

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/portletlist/PortletListRESTController.java [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    public @ResponseBody String updatePortletList(
+        HttpServletRequest request,
+        HttpServletResponse response,
+        @RequestBody String json,
+        @PathVariable String portletListUuid) {
+
+        final IPerson person = personManager.getPerson(request);
+        if(log.isDebugEnabled()) {
+            debugPerson("updatePortletList", person);
+            log.debug("updatePortletList > JSON body is = {}", json);
```
[the remark this answers is not in the document]
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/portletlist/PortletListRESTController.java:187]: Seems reasonable. I'll fix these

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/dao/portletlist/jpa/PortletListItem.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    @JsonIgnore
+    @Id
+    @GeneratedValue(generator = "UUID")
+    @GenericGenerator(
+        name = "UUID",
+        strategy = "org.hibernate.id.UUIDGenerator"
+    )
+    @Column(name = "ID", updatable = false, nullable = false)
+    private String id;
```
[the remark this answers is not in the document]
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/dao/portletlist/jpa/PortletListItem.java:41]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/dao/portletlist/jpa/PortletListItem.java:41]: I think it's getting hung up on the annotations. Doesn't look like an issue

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/portletlist/PortletListRESTController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/portletlist/PortletListRESTController.java]: I've removed the logging statement(s).

File: uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/dao/portletlist/jpa/PortletListItem.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import lombok.Setter;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.apereo.portal.dao.portletlist.IPortletListItem;
+import org.hibernate.annotations.CacheConcurrencyStrategy;
+import org.hibernate.annotations.GenericGenerator;
+
+@Getter
+@Setter
+@EqualsAndHashCode
```
[the remark this answers is not in the document]
Comment (cbeach47) [reply uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/dao/portletlist/jpa/PortletListItem.java:16]: not relevant - closing.

**Comment by cbeach47** [comment]
hardened the logging messages as per Sonatype.  I think the rest are OK to ignore for now, basic logging of exceptions / stack traces, or existing issues.

**Comment by bjagg** [comment]
For expediency, I am going to merge and then remove the duplicate class in a new PR.

**Pull request merged**